### PR TITLE
Initial version of metacp

### PR DESCRIPTION
**Pull request by xeno-by**
Some functionality is missing, but it's already useful.

### CODE REVIEW

File: tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala [hunk 1]
```diff
@@ -50,12 +55,29 @@ trait ExpectHelpers {
...
+    val classpath = BuildInfo.databaseClasspath
+    val target = Files.createTempDirectory("target_")
+    val metacp_settings = MetacpSettings.parse(List("-d", target.toString, classpath)).get
+    val (metacp_exitcode, _) = CliSuite.communicate(Metacp.process(metacp_settings))
+    assert(metacp_exitcode == 0)
+    val paths = Files.walk(target).iterator.asScala
+    val semanticdbs = paths.filter(_.toFile.isFile).map(_.toString).toArray
+    val (metap_exitcode, metap_stdout) = CliSuite.communicate(Metap.process(semanticdbs))
+    assert(metap_exitcode == 0)
+    metap_stdout
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala]: Alternatively, we can add another `metacp-metap-diff.expect` that is the unified diff between the two for the for the symbols that metacp converts. In theory, that file should be empty but if there are known differences between the two then it will be highlighted here.

File: semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala [hunk 2]
```diff
@@ -1,11 +1,209 @@
...
+  }
+
+  def process(settings: Settings): Int = {
+    var failed = false
+    val classpath = Classpath(settings.cps.mkString(File.pathSeparator))
+    val fragments = classpath.deep.filter(_.uri.toString.endsWith(".class"))
+    fragments.sortBy(_.uri.toString).foreach { fragment =>
+      try {
+        val bytecode = {
+          val is = fragment.uri.toURL.openStream()
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala]: Done. Forgot that this method exists.

File: semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala [hunk 3]
```diff
@@ -1,11 +1,209 @@
...
+          val os = new ByteArrayOutputStream()
+          var b = is.read()
+          while (b != -1) {
+            os.write(b)
+            b = is.read()
+          }
+          is.close()
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala]: Done.

File: semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala [hunk 4]
```diff
@@ -1,11 +1,209 @@
...
+            val semanticdbDocument = s.TextDocument(
+              schema = s.Schema.SEMANTICDB3,
+              uri = className.replace(".class", ".scala"),
+              language = "Scala",
+              symbols = semanticdbInfos)
+            val semanticdbDocuments = s.TextDocuments(List(semanticdbDocument))
+            val semanticdbStream = Files.newOutputStream(semanticdbFile)
+            semanticdbDocuments.writeTo(semanticdbStream)
+            semanticdbStream.close()
+          case None =>
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala]: Done.

File: semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala [hunk 5]
```diff
@@ -1,11 +1,209 @@
...
+        val classfile = ClassFileParser.parse(bytecode)
+        ScalaSigParser.parse(classfile) match {
+          case Some(scalaSig) =>
+            val semanticdbInfos = scalaSig.symbols.flatMap {
+              case sym: SymbolInfoSymbol => sinfo(sym)
+              case _ => None
+            }
+            val semanticdbRoot = Paths.get(settings.d, "META-INF", "semanticdb")
+            val className = NameTransformer.decode(fragment.name.toString)
+            val semanticdbName = className.replace(".class", ".semanticdb")
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala]: Wanted to do this for ages - thanks for bringing this up. I'll also change `semanticdb-scalac`.

File: semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala [hunk 6]
```diff
@@ -1,11 +1,209 @@
...
+            val semanticdbFile = semanticdbRoot.resolve(semanticdbName)
+            semanticdbFile.getParent.toFile.mkdirs()
+            val semanticdbDocument = s.TextDocument(
+              schema = s.Schema.SEMANTICDB3,
+              uri = className.replace(".class", ".scala"),
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala]: Indeed.

File: tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala [hunk 7]
```diff
@@ -24,6 +25,10 @@ class ExpectSuite extends FunSuite with DiffAssertions {
     // output of 2.12. It's possible to add another expect file for 2.12
     // later down the road if that turns out to be useful.
     case "2" :: "12" :: Nil =>
+      test("convert.expect") {
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala]: Done.

File: semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala [hunk 8]
```diff
@@ -1,11 +1,209 @@
...
+              case _ => "<?>"
+            }
+          }
+          loop(sym.infoType)
+        }
+        prefix + encodedName + descriptor + "."
+      case k.TYPE | k.CLASS | k.TRAIT =>
+        prefix + encodedName + "#"
+      case k.PARAMETER =>
+        prefix + "(" + encodedName + ")"
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala]: To tell the truth, I've been quite burned by all the abstractions that we built around SemanticDB v1/v2, and now I'm enjoying bare metal stuff. I suggest we let things evolve for the time being, and then deduplicate / abstract as appropriate once we see the patterns emerge.

File: semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala [hunk 9]
```diff
@@ -193,17 +206,192 @@ object Main {
...
+          Some(s.Type(tag = stag, methodType = Some(s.MethodType(Nil, sparamss, sret))))
+        case TypeBoundsType(lo, hi) =>
+          val stag = t.TYPE_TYPE
+          val slo = loop(lo)
+          val shi = loop(hi)
+          Some(s.Type(tag = stag, typeType = Some(s.TypeType(Nil, slo, shi))))
+        case PolyType(tpe, tparams) =>
+          val stparams = tparams.map(ssymbol)
+          val stpe = loop(tpe)
+          stpe.map { stpe =>
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala]: Refactored.

File: semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala [hunk 10]
```diff
@@ -1,11 +1,209 @@
...
+          case Some(scalaSig) =>
+            val semanticdbInfos = scalaSig.symbols.flatMap {
+              case sym: SymbolInfoSymbol => sinfo(sym)
+              case _ => None
+            }
+            val semanticdbRoot = Paths.get(settings.d, "META-INF", "semanticdb")
+            val className = NameTransformer.decode(fragment.name.toString)
+            val semanticdbName = className.replace(".class", ".semanticdb")
+            val semanticdbFile = semanticdbRoot.resolve(semanticdbName)
+            semanticdbFile.getParent.toFile.mkdirs()
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala]: Nice catch! This is what I meant in another comment by deduplicating once patterns start to emerge. Will do.

File: tests/jvm/src/test/scala/scala/meta/tests/cli/CliSuite.scala [hunk 11]
```diff
@@ -20,7 +20,7 @@ class CliSuite extends FunSuite with DiffAssertions {
     }
   """.getBytes(UTF_8))
   val target = Files.createTempDirectory("target_")
-  val helloWorldSemanticdb = target.resolve("META-INF/semanticdb/HelloWorld.semanticdb")
+  val helloWorldSemanticdb = target.resolve("META-INF/semanticdb/HelloWorld.scala.semanticdb")
```
Comment (xeno-by) [inline tests/jvm/src/test/scala/scala/meta/tests/cli/CliSuite.scala]: @olafurpg This is potentially controversial, but we don't provide any [compatibility guarantees](https://github.com/scalameta/scalameta/blob/master/COMPATIBILITY.md) for the file layout.

Comment (xeno-by) [reply tests/jvm/src/test/scala/scala/meta/tests/cli/CliSuite.scala]: Review of our tools:
  * `semanticdb-scala` [already does the same](https://github.com/scalameta/semanticdb-sbt/blob/master/sbthost/nsc/src/main/scala/scala/meta/internal/sbthost/ConfigOps.scala#L10).
  * `Scalafix` [uses Database.load](https://github.com/scalacenter/scalafix/blob/master/scalafix-cli/src/main/scala/scalafix/cli/CliRunner.scala#L345), which in turn [calls v.Database.toSchema](https://github.com/scalameta/scalameta/blob/master/langmeta/langmeta/shared/src/main/scala/org/langmeta/semanticdb/Database.scala#L47), which [doesn't care about on-disk filenames](https://github.com/scalameta/scalameta/blob/master/langmeta/langmeta/shared/src/main/scala/org/langmeta/semanticdb/internal/vfs/Database.scala#L22) as long as [they end with "semanticdb"](https://github.com/scalameta/scalameta/blob/master/langmeta/langmeta/shared/src/main/scala/org/langmeta/semanticdb/internal/vfs/SemanticdbPaths.scala#L13).
  * Metadoc already does the same (multiple links don't fit here).
  * Metals doesn't care about filenames at all.

**Comment by xeno-by** [comment]
@olafurpg Addressed all your comments.

File: tests/jvm/src/test/scala/scala/meta/tests/cli/CliSuite.scala (hunk 11 above)
Comment (olafurpg) [reply tests/jvm/src/test/scala/scala/meta/tests/cli/CliSuite.scala]: Just do it. I don't rely on the filenames because the URI is embedded in the proto message

File: langmeta/langmeta/shared/src/main/scala/org/langmeta/io/AbsolutePath.scala [hunk 12]
```diff
@@ -29,6 +30,7 @@ sealed abstract case class AbsolutePath(toNIO: nio.Path) {
   def isFile: Boolean = FileIO.isFile(this)
   def isDirectory: Boolean = FileIO.isDirectory(this)
   def readAllBytes: Array[Byte] = FileIO.readAllBytes(this)
+  def write(proto: GeneratedMessage): Unit = FileIO.write(this, proto)
```
Comment (olafurpg) [inline langmeta/langmeta/shared/src/main/scala/org/langmeta/io/AbsolutePath.scala]: Let's not add semanticdb specific methods to `AbsolutePath`. In fact, I regret adding `readAllBytes/isDirectory/isFile`, I think those should be static methods on some object like `nio.Files`. I would love to tear out org.langmeta.io into a standalone project one day so this method here is an odd fish.

File: tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala [hunk 13]
```diff
@@ -25,8 +25,8 @@ class ExpectSuite extends FunSuite with DiffAssertions {
     // output of 2.12. It's possible to add another expect file for 2.12
     // later down the road if that turns out to be useful.
     case "2" :: "12" :: Nil =>
-      test("convert.expect") {
-        import ConvertExpect._
+      test("metacp.expect") {
```
Comment (olafurpg) [inline tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala]: I feel like we should test the diff between metacp and metac for symbols that are covered by metacp. In theory they should produce identical types making the diff empty but I suspect that may not be the case, if there are inconsistencies I want them documented in our tests.

File: semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala [hunk 14]
```diff
@@ -46,7 +46,7 @@ object Main {
               case _ => None
             }
             val semanticdbRoot = Paths.get(settings.d, "META-INF", "semanticdb")
-            val className = NameTransformer.decode(fragment.name.toString)
+            val className = NameTransformer.decode(fragment.name.toString.replace("\\", "/"))
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala]: Why this change? Seems ad-hoc. Is this supposed to be `PathIO.toUnix`?

File: langmeta/langmeta/shared/src/main/scala/org/langmeta/io/AbsolutePath.scala (hunk 12 above)
Comment (xeno-by) [reply langmeta/langmeta/shared/src/main/scala/org/langmeta/io/AbsolutePath.scala]: Done.

File: semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala (hunk 14 above)
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/cli/metacp/Main.scala]: Done.

File: tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala (hunk 13 above)
Comment (xeno-by) [reply tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala]: Let's do this in a different pull request, since this one is getting pretty big. This is going to be my third rebase, and keeping this PR in the queue for longer will mean more rebases.

File: langmeta/langmeta/shared/src/main/scala/org/langmeta/io/AbsolutePath.scala (hunk 12 above)
Comment (olafurpg) [reply langmeta/langmeta/shared/src/main/scala/org/langmeta/io/AbsolutePath.scala]: Remove this method
